torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -0,0 +1,485 @@
+
+        Arguments:
+            closure (callable): A closure that reevaluates the model and
+                returns the loss. Optional for most optimizers.
+        Returns:
+            optional loss, depends on the underlying optimizer
+
+        .. note: Any extra parameter is passed to the base optimizer as-is"""
+
+        # Sync oss param_groups attributes in case they've been updated by a scheduler.
+        self._sync_param_groups()
+
+        # Run the optimizer step on this shard only:
+        if closure is not None:
+            loss = self.optim.step(closure=closure, **kwargs)  # type: ignore
+        else:
+            loss = self.optim.step(**kwargs)
+
+        # Sync all the updated shards in between the ranks
+        self._broadcast_params()
+
+        # Sync hypothethical new results from the wrapped optimizer to the exposed param_groups
+        self._sync_param_groups(local_to_global=True)
+
+        return loss
+
+    def load_local_state_dict(self, state_dict: dict) -> None:
+        """Loads this rank's state_dict.
+
+        .. warning: This is not meant to load the global state dict.
+        """
+
+        self.optim.load_state_dict(state_dict)
+
+        # Workaround PyTorch bug that casts state (https://github.com/pytorch/pytorch/issues/43706)
+        # Copied from https://github.com/pytorch/fairseq/blob/v0.9.0/fairseq/optim/fp16_optimizer.py#L251-L268
+        groups = self.optim.param_groups
+        saved_groups = state_dict["param_groups"]
+        id_map = {
+            old_id: p
+            for old_id, p in zip(chain(*(g["params"] for g in saved_groups)), chain(*(g["params"] for g in groups)))
+        }
+        for k, v in state_dict["state"].items():
+            if k in id_map:
+                param = id_map[k]
+                self.optim.state[param] = _recursive_copy_to_device(v, non_blocking=True, device=param.device)
+
+        # Restore the global param_groups (the params themselves are already correct)
+        for global_group, local_group in zip(self.param_groups, groups):
+            for k, v in local_group.items():
+                if k != "params":
+                    global_group[k] = v
+
+        # Force a re-partitioning, in case the model changed with the new state
+        self._partition_parameters.clear()
+        self._per_device_params.clear()
+        self._param_rank.clear()
+
+        # Update the bucketing strategy accordingly
+        self._setup_bucket_strategy()
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Restore the global parameter groups as well as the shard.
+
+        Arguments:
+            state_dict (dict): optimizer state. Should be an object returned
+                from a call to :meth:`state_dict`
+        """
+
+        # Check whether we got a local or global dict
+        if "local_state_dict" in state_dict and state_dict["local_state_dict"]:
+            self.load_local_state_dict(state_dict)
+        else:
```

Comment:
trying to verify that my understanding is correct. If `consolidate_state_dict` has been call with `recipient_rank=x`, then x's `state_dict()` will return the global one, and when calling `load_state_dict()` on x, it will go through the `else` branch?

For all other ranks, their `state_dict()` would only return local states, and when calling `load_state_dict`, it will go through the if branch here?

If this will be the long term solution, it will be very helpful to have an example in the doc showing that.

Reply:
> trying to verify that my understanding is correct. If `consolidate_state_dict` has been call with `recipient_rank=x`, then x's `state_dict()` will return the global one, and when calling `load_state_dict()` on x, it will go through the `else` branch?

correct with this version, but I think that the change that we discussed (return a pure pytorch state, https://github.com/facebookresearch/fairscale/pull/310) is strictly superior, it removes this branch basically

> For all other ranks, their `state_dict()` would only return local states, and when calling `load_state_dict`, it will go through the if branch here?

it was a somewhat weird-but-why-not possibility, open for users to check in and out the sharded state instead of the global one basically, but I don't think that anyone actually used that

> If this will be the long term solution, it will be very helpful to have an example in the doc showing that.

I think that it's better to change that to stick to a (monolithic) pytorch state, if anything that's one less difference with a normal optimizer to learn, it lowers the bar and that's often a good thing